### PR TITLE
Inject all metadata at once after the action is executed.

### PR DESCRIPTION
**Pull request by coeuvre**
So that spawns within one action will be able to change the injected metadata using e.g. renameTo.

Also only allows metadata injection for declared action outputs instead of declared spawn outputs.

Related #16351.

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java [hunk 1]
```diff
@@ -84,27 +90,39 @@ public void updateContext(MetadataInjector metadataInjector) {
...
   void injectFile(ActionInput file, RemoteFileArtifactValue metadata) {
-    checkNotNull(metadataInjector, "metadataInject is null");
-
     injectedMetadata.put(file.getExecPath(), metadata);
+  }
 
-    if (file instanceof Artifact) {
-      metadataInjector.injectFile((Artifact) file, metadata);
+  void flush() {
+    checkNotNull(metadataInjector, "metadataInject is null");
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java:110]:
```suggestion
    checkNotNull(metadataInjector, "metadataInjector is null");
```

File: src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java [hunk 2]
```diff
@@ -84,27 +90,39 @@ public void updateContext(MetadataInjector metadataInjector) {
...
+
+    for (Map.Entry<PathFragment, Artifact> entry : outputMapping.entrySet()) {
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java:112]: This seems to be a functional change, as previously `injectFile` and `injectTree` would perform the injection even when the artifact was not an output. Is this intentional? (I'm not aware of a reason to inject non-outputs, but that doesn't mean there isn't one :))

Comment (coeuvre) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java:112]: It is intentional. The difference is, before this change, we could only inject declared spawn outputs (not arbitrary output) which is a super set of declared action outputs. In most cases they should be same (i.e one action generates one spawn). In the cases that declared spawn outputs have more outputs than declared action outputs, most of them are instance of `ActionInput` which don't get injected.

From the perspective of reading the injected metadata, all that matters is metadata for declared outputs. I can't find a reason for allowing inject metadata of non-declared outputs (buy I may be wrong).

File: src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java (hunk 1 above)
Comment (coeuvre) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java]: Done.

**Review by tjgq — APPROVED** [review]

**Pull request closed without merging**